prov/util/src/util_mem_hooks.c

```diff
@@ -362,7 +405,14 @@ static int ofi_intercept_symbol(struct ofi_intercept *intercept)
 
 	intercept->orig_func = func_addr;
 
-	ret = ofi_patch_function(intercept);
+	if (ofi_is_function_patched(intercept)) {
+		FI_WARN(&core_prov, FI_LOG_MR,
+				"function %s is already patched; stopping further patching\n",
+				intercept->symbol);
+		ret = -EAGAIN;
```

Comment:
I do not think EAGAIN is the right error code here. Maybe -FI_EALREADY?

Reply:
I'd be ok with -FI_EALREADY.

Reply:
Also, I have a question that may be out of the scope this commit:

how does the caller handle the error? e.g. would the application fail, or the application continue with the MR cache turned off,  or the application continue with a different provider that does not need MR cache?

Reply:
I've made the change to -FI_EALREADY, thank you for the feedback.

I will let Brian answer the other question as we were both talking about that earlier, and that is something he is working on. For now, the scope here is to at least detect this condition.

Reply:
I see. Thank you!

Reply:
Today, the call to `ofi_mr_cache_init()` will fail.  That's probably not what we want, given that will break Open MPI without further work.  We're going to work at that part, so this becomes somewhat of an ordering exercise.

One question for @shefty is how we want to tackle handling this failure during initialization.  Today, the first monitor code to run that returns an error code is `monitor->start()` as part of `ofi_mr_cache_init()`.  That seems a bit late to be swapping monitors on a provider.  Individual providers can handle this exception explicitly.  EFA could call `monitor->start()` when initializing the domain, and adjust the monitor to uffd on failure.  It seems like we could handle that "better" for some definition of better, but maybe not.

Because of release timings with Open MPI, we probably don't want to run the already patched check until domains are opened (when we install the patcher today), since otherwise Open MPI will just overwrite Libfabric's patcher.